File: organism/haskell/src/Organism/Routing.hs
-- | Pure functional phi-weighted routing over the protocol mesh.
module Organism.Routing
  ( Task(..), ScoredRoute(..), route, topRoutes, applyFeedback
  , RoutingTable, buildRoutingTable
  ) where

import Data.List (sortBy)
import Data.Ord  (comparing, Down(..))
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Organism.Constants (phi, phiInv)
import Organism.Protocol  (ProtocolSpec(..), ProtocolId(..), RingAffinity)

data Task = Task
  { taskKeyword   :: String
  , taskRing      :: Maybe RingAffinity
  , taskCapability :: Maybe String
  } deriving (Show)

data ScoredRoute = ScoredRoute
  { routeSpec  :: ProtocolSpec
  , routeScore :: Double
  } deriving (Show)

-- | Routing table: maps protocol id to (spec, adaptive score).
type RoutingTable = Map ProtocolId (ProtocolSpec, Double)

-- | Build initial routing table from a list of specs.
-- Base score = φ^(-i/n) where i is ordinal position.
buildRoutingTable :: [ProtocolSpec] -> RoutingTable
buildRoutingTable specs =
  let n = fromIntegral (length specs)
      indexed = zip [0..] specs
  in  Map.fromList
        [ (specId s, (s, phi ** (-(fromIntegral i) / n)))
        | (i, s) <- indexed
        ]

-- | Score a single protocol spec against a task.
score :: RoutingTable -> Task -> ProtocolSpec -> Double
score table task spec =
  let base = maybe 1.0 snd (Map.lookup (specId spec) table)
      ringBonus = case taskRing task of
        Just r | r == specRing spec -> phi * phi
        _                           -> 1.0
      capBonus = case taskCapability task of
        Just c | c `elem` specCapabilities spec -> phi
        _                                       -> 1.0
      kwBonus | taskKeyword task `elem` specCapabilities spec = phi
              | otherwise                                     = 1.0
  in  base * ringBonus * capBonus * kwBonus

-- | Find the best-matching protocol for a task.
route :: RoutingTable -> [ProtocolSpec] -> Task -> Maybe ScoredRoute
route _ [] _ = Nothing
route table specs task =
  let scored = [ ScoredRoute s (score table task s) | s <- specs ]
      best   = head (sortBy (comparing (Down . routeScore)) scored)
  in  Just best

-- | Top-n routes for a task.
topRoutes :: RoutingTable -> [ProtocolSpec] -> Task -> Int -> [ScoredRoute]
topRoutes table specs task n =
  take n . sortBy (comparing (Down . routeScore))
  $ [ ScoredRoute s (score table task s) | s <- specs ]

-- | Phi-decay adaptive feedback — updates score in routing table.
applyFeedback :: RoutingTable -> ProtocolId -> Bool -> RoutingTable
applyFeedback table pid success =
  Map.adjust update pid table
  where
    update (spec, s)
      | success   = (spec, min (phi^(3::Int)) (s * phi))
      | otherwise = (spec, max (phi**(-3))    (s * phiInv))

File: organism/julia/src/Routing.jl
"""
    Organism.Routing

Phi-weighted intelligent routing over the protocol mesh.
Routes tasks to optimal protocols using adaptive phi-decay scoring.
"""
module Routing

using ..Constants: PHI, PHI_INV

export Route, Router, route, register_outcome!, top_routes

struct Route
    protocol_id::String
    name::String
    ring::String
    intelligence_class::String
    score::Float64
end

mutable struct Router
    protocols::Vector{Dict{String,String}}
    scores::Dict{String,Float64}
end

function Router(protocols::Vector{Dict{String,String}})
    n = length(protocols)
    scores = Dict{String,Float64}()
    for (i, p) in enumerate(protocols)
        # Base score: phi^(-i/n) — earlier protocols higher priority
        scores[get(p, "protocol_id", string(i))] = PHI ^ (-(i-1) / max(1, n))
    end
    Router(protocols, scores)
end

"""
    route(r, task) -> Route

Find the highest-scoring protocol for a task. Task is a NamedTuple or Dict
with optional fields: ring, intelligence_class, keyword.
"""
function route(r::Router, task::NamedTuple = NamedTuple())
    best = nothing
    best_score = -Inf
    ring_kw = lowercase(get(task, :ring, ""))
    cls_kw  = lowercase(get(task, :intelligence_class, ""))
    kw      = lowercase(get(task, :keyword, ""))

    for p in r.protocols
        pid = get(p, "protocol_id", "")
        score = get(r.scores, pid, 1.0)

        ring = lowercase(get(p, "ring_affinity", ""))
        cls  = lowercase(get(p, "intelligence_class", ""))
        text = lowercase(join([get(p, "protocol_name", ""), get(p, "primary_function", "")], " "))

        !isempty(ring_kw) && occursin(ring_kw, ring) && (score *= PHI^2)
        !isempty(cls_kw)  && occursin(cls_kw, cls)   && (score *= PHI)
        !isempty(kw)      && occursin(kw, text)       && (score *= PHI)

        if score > best_score
            best_score = score
            best = Route(pid, get(p, "protocol_name", ""), get(p, "ring_affinity", ""),
                         get(p, "intelligence_class", ""), score)
        end
    end
    best
end

"""
    register_outcome!(r, protocol_id, success)

Phi-decay adaptive feedback. Successful routes amplify by φ, failures decay.
"""
function register_outcome!(r::Router, protocol_id::String, success::Bool)
    prior = get(r.scores, protocol_id, 1.0)
    r.scores[protocol_id] = success ? min(PHI^3, prior * PHI) : max(PHI^-3, prior * PHI_INV)
end

"""
    top_routes(r, task, n) -> Vector{Route}

Return top-n protocol routes for a task, sorted by score descending.
"""
function top_routes(r::Router, task::NamedTuple = NamedTuple(), n::Int = 5)
    routes = Route[]
    for p in r.protocols
        pid = get(p, "protocol_id", "")
        score = get(r.scores, pid, 1.0)
        ring = lowercase(get(p, "ring_affinity", ""))
        cls  = lowercase(get(p, "intelligence_class", ""))
        text = lowercase(join([get(p, "protocol_name", ""), get(p, "primary_function", "")], " "))
        ring_kw = lowercase(get(task, :ring, ""))
        cls_kw  = lowercase(get(task, :intelligence_class, ""))
        kw      = lowercase(get(task, :keyword, ""))
        !isempty(ring_kw) && occursin(ring_kw, ring) && (score *= PHI^2)
        !isempty(cls_kw)  && occursin(cls_kw, cls)   && (score *= PHI)
        !isempty(kw)      && occursin(kw, text)       && (score *= PHI)
        push!(routes, Route(pid, get(p, "protocol_name", ""), ring,
                            get(p, "intelligence_class", ""), score))
    end
    sort!(routes, by = r -> r.score, rev = true)
    routes[1:min(n, length(routes))]
end

end # module
